packages/node/src/signing-keys-generator.ts

```diff
@@ -0,0 +1,67 @@
+import { Node } from "@counterfactual/types";
+import { Wallet } from "ethers";
+import { hashMessage } from "ethers/utils";
+import { Memoize } from "typescript-memoize";
+
+export class SigningKeysGenerator {
+  // The namespacing is on the channel the AppInstance is in
+  private namespacedAppInstanceNumberToSigningKey: Map<
+    string,
+    string
+  > = new Map();
+  private signingKeys: Set<string> = new Set();
+
+  constructor(
+    private readonly privateKeyGenerator: Node.IPrivateKeyGenerator
+  ) {}
+
+  @Memoize()
+  /**
+   * @param appSequenceNumber
+   * @param multisigAddress The namespace under which `appSequenceNumber` is
+   *        used since the provided callback function is idempotent for any
+   *        value passed in. This prevents collisions that'd occur on  using the
+   *        same `appSequenceNumber` across different channels from receiving the
+   *        same signing key from the callback.
+   */
+  public async getSigningKey(
+    appSequenceNumber: number,
+    multisigAddress: string
+  ): Promise<string> {
+    const appInstanceUniqueId = hashMessage(
+      `${multisigAddress}_${appSequenceNumber}`
+    );
+
+    if (this.namespacedAppInstanceNumberToSigningKey.has(appInstanceUniqueId)) {
+      return await this.namespacedAppInstanceNumberToSigningKey.get(
+        appInstanceUniqueId
+      )!;
+    }
+
+    // FIXME: fallback on mnemonic-based signer if callback isn't specified
+    // in which case the derivation path needs to be converted into a decimal
+    // representation for it to be a valid path
+    const signingKey = await this.privateKeyGenerator(appInstanceUniqueId);
+    try {
+      new Wallet(signingKey);
+    } catch (e) {
```

Comment:
I'm guessing `ethers` throws some very specific types of errors here, maybe catch those?